torch/csrc/jit/serialization/export_module.cpp

```diff
@@ -190,9 +179,35 @@ std::pair<IValue, IValue> getFunctionTuple(
   types.reserve(code->type_table().size());
   static const std::string torch_prefix("__torch__");
   static const std::string class_prefix("__torch__.torch.classes");
+  std::shared_ptr<torch::jit::CompilationUnit> cu =
+      module._ivalue()->compilation_unit();
   for (const TypePtr& t : code->type_table()) {
     auto type_str = t->annotation_str();
-    if (type_str.find(torch_prefix) == 0) {
+    if (cu->get_named_tuple(t->str())) {
+      auto named_tuple_type = cu->get_named_tuple(t->str());
+      if (named_tuple_type != nullptr) {
+        IValue named_tuple{"NamedTuple"};
+        std::vector<IValue> name_type_pairs;
+        // Get the field name and field type for the NamedTuple
+        for (auto const& argument : named_tuple_type->schema()->arguments()) {
+          name_type_pairs.emplace_back(c10::ivalue::Tuple::create(
+              {argument.name(), argument.type()->repr_str()}));
+        }
+        // Create a tuple with following structure
+        // ('type_name',
+        //   ('NamedTuple,
+        //     ('filed_name_1', 'field_type_1')
+        //     ('filed_name_2', 'field_type_2')
+        //   )
+        // )
+        IValue named_tuple_type_definition =
+            c10::ivalue::Tuple::create(std::vector<IValue>{
+                named_tuple, c10::ivalue::Tuple::create(name_type_pairs)});
+        types.emplace_back(c10::ivalue::Tuple::create(std::vector<IValue>{
+            IValue(t->str()), named_tuple_type_definition}));
```

Comment:
Do we need to use the mangled (unique) name of the type of the NamedTuple `__torch__. ...` by using  `type_name_uniquer_.getUniqueName` like we do for modules.

For reference please check this. Previously when we have sub modules that have same name, the names were not not mangled hence in `bytecode.pkl` there was only one reference.
https://github.com/pytorch/pytorch/pull/61933

as @iseeyuan mentioned, this might not be required for NamedTuple objects. Please correct if I am wrong, to check this, we have multiple instance of same NamedTuple object like,

```
self.foo_1 = Foo(torch.tensor(1))
self.foo_2 = Foo(torch.tensor(1))
...
```

to see if it gets saved and loaded as expected.

Reply:
Thanks for providing the example code. I check this case, and it looks fine. The source code I used is 
```
class Foo(NamedTuple):
    id: torch.Tensor

class Bar(torch.nn.Module):
    def __init__(self):
        super(Bar, self).__init__()
        # self.foo = Foo(torch.tensor(1))
        self.foo_1 = Foo(torch.tensor(1))
        self.foo_2 = Foo(torch.tensor(1))

    def forward(self, a: torch.Tensor):
        self.foo_1 = Foo(a)
        return (self.foo_1, self.foo_2)

```

and the corresponding TorchScript code is: 
```
class Bar(Module):
  __parameters__ = []
  __buffers__ = []
  training : bool
  _is_full_backward_hook : Optional[bool]
  foo_1 : Tuple[Tensor]
  foo_2 : Tuple[Tensor]
  def forward(self: __torch__.Bar,
    a: Tensor) -> Tuple[Tuple[Tensor], Tuple[Tensor]]:
    self.foo_1 = __torch__.Foo(a,)
    self.foo_2 = __torch__.Foo(a,)
    foo_1 = self.foo_1
    foo_2 = self.foo_2
    return (foo_1, foo_2)
class Foo(NamedTuple):
  id : Tensor
```
After thinking about it, each type name in the type table is associated with TorchScript source code (`__torch__.Foo` in the example case). This name should be unique, because class definition in the same python file under the same corresponding path is unique. For example, the following code will only have one valid `Foo` class definition in
`test.py`
```
class Foo(NamedTuple):
    id: torch.Tensor

class Foo(NamedTuple):
    id: torch.Tensor
    age: torch.Tensor
```

Reply:
thanks for checking this and clarifying it 👍